Fix Kernel.trap setup for sigint by going from a lambda to a Proc

starbelly opened the pull request:
It should be noted that a better test can be put in place, but figured I do that in another PR. 
Specifically, we probably want send the process a ctrl-c sequence by writing to stdin, and verifying we get our warning, etc. But maybe that should be part of this PR as well. Wasn't sure how hacky the test would be, might have to use expect. 

 - Switched out the block passed to Kernel.trap for SIGINT from a lambda
to proc. The block given to Kernel.trap gets a signal argument. Not
expecting this argument in a lambda caused a runtime error and prevented graceful exiting
- Added a few test units using popen3 to make sure the process gets
  sucessfully killed when a sigint is sent
- Add a test helper spawn_bin to crank up shindo in a monitorable way
  via Open3::popen3
- Add bin_is_alive? Helper to check an arbitrary pid

starbelly commented:
One of the builds passed (2.0.0). jruby fails because wait_thr.pid is returning nil, not sure why, not too familar with jruby and caveats thereof. All the other builds fail become of a gem dependency/conflict regarding rake.

geemus commented:
Sounds good (and looks good so far at least). Just let me know if there is anything I can help with. I'm guessing we could lock to an earlier version of rake and get those old-rubies working, but alternatively we could probably reasonably consider dropping support for some or all of them at this point (or at least soon).

starbelly commented:
@geemus I'm in favor of not supporting the older rubies, but not sure how many people out there are using shindo on old rubies. With that said, what's your preference. Also, I've got better tests for MRI, but jruby... "Cutty say 'e can't HANG!" : ) Yeah, I can't think of solid way to test this with jruby. It seems that by default, ctrl-c during a shindo session is not going work anyway ( need to install a signal handler to emulate SIGINT, which I'll do because it's needed anyway). Though, the bigger problem is sending sigint to a thread. The only way I can see to do this is to send use expect, but I don't like that external dependency.  I haven't tried pty yet, maybe that would work. I obviously have no real experience with jruby, so if you or anyone else can offer some insight that would be great (*lumberg voice*). Of course, it might worth it just to skip this test on jruby. 

"I have people skills; I am good at dealing with people. Can't you understand that? What the hell is wrong with you people?!"

starbelly commented:
@geemus Now all the MRI test are failing : ) Jruby passes though, but we're not testing it : p 

 A few builds failed because of gemspec error, the other MRI builds I'm not sure. Those rubies all pass on my machine, so let me try them on a linux server and see what I get.

 On the rake dependency note, I did add in a bit to lock rubies >= 1.8 to rake 10.5. Though,  I think we should definitely drop the support for rubies older than 2.0.0, maybe not even 2.0.0, as it's no longer maintained.

geemus commented:
Yeah, I think it is probably not unreasonable for us to start dropping unsupported rubies. We have a version they can use, we just won't be making new releases (it's been fairly stable for quite a while anyway). Similarly, I don't know of that much jruby usage, so in a pinch we can probably move that to allowed failures. It would be good to get the supported MRIs passing again though, I wonder why they stopped...

starbelly commented:
@geemus Agreed. I need to get the tests to work with MRI again, which they do, just not within the build. Anyway, not done here, just put it down for a minute to do some other work. Will change title to WIP.

geemus commented:
Sounds good, thanks!

On Sat, Jun 11, 2016 at 2:19 PM, Bryan Paxton <notifications@github.com>
wrote:

> @geemus <https://github.com/geemus> Agreed. I need to get the tests to
> work with MRI again, which they do, just not within the build. Anyway, not
> done here, just put it down for a minute to do some other work. Will change
> title to WIP.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/geemus/shindo/pull/16#issuecomment-225388544>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AAAQKugABxPzxj1Drwq0qN0CCoSFw6vEks5qKwpUgaJpZM4Ixn8Q>
> .
>

starbelly commented:
@geemus This can be merged now. We are forgoing the test units with the notion of Shindo being legacy and with intention to replace Shindo with Rspec in excon, fog, etc. None the less,  this PR does fix the issue at hand. I also removed the two oldest rubies and added 2.3.1 to travis.yml

geemus commented:
Thanks!
